tools/clang/lib/SPIRV/EmitSPIRVAction.cpp

```diff
@@ -372,12 +374,98 @@ class SPIRVEmitter : public ASTConsumer {
     theBuilder.setInsertPoint(mergeBB);
   }
 
+  void doForStmt(const ForStmt *forStmt) {
+    // for loops are composed of:
+    //   for (<init>; <check>; <continue>) <body>
+    //
+    // To translate a for loop, we'll need to emit all <init> statements
+    // in the current basic block, and then have separate basic blocks for
+    // <check>, <continue>, and <body>. Besides, since SPIR-V requires
+    // structured control flow, we need two more basic blocks, <header>
+    // and <merge>. <header> is the block before control flow diverges,
+    // while <merge> is the block where control flow subsequently converges.
+    // The final CFG should normally be like the following. Exceptions will
+    // occur with non-local exits like loop breaks or early returns.
+    //             +--------+
+    //             |  init  |
+    //             +--------+
+    //                 |
+    //                 v
+    //             +--------+
+    //             | header | <----------------+
+    //             +--------+                  |
```

Comment:
Good suggestion! That's true. Was literally following the spec. Actually this is a good optimization. Done.